File: Range/range_processing.cpp
#include "range_processing.hpp"
#include <cmath>
#include <algorithm>
#include <fftw3.h>

vector<double> range_processing(const vector<complex<double>>& rx_signal, int num_samples) {
    // Apply Hann window
    vector<complex<double>> windowed_signal(num_samples);
    for (int i = 0; i < num_samples; ++i) {
        double hann = 0.5 * (1 - cos(2 * M_PI * i / (num_samples - 1)));
        windowed_signal[i] = hann * rx_signal[i];
    }
    
    // Prepare for FFT
    fftw_complex *in, *out;
    fftw_plan p;
    
    in = (fftw_complex*) fftw_malloc(sizeof(fftw_complex) * num_samples);
    out = (fftw_complex*) fftw_malloc(sizeof(fftw_complex) * num_samples);

    // Copy windowed signal to FFTW input
    for (int i = 0; i < num_samples; ++i) {
        in[i][0] = windowed_signal[i].real();
        in[i][1] = windowed_signal[i].imag();
    }

    // Create FFT plan and execute
    p = fftw_plan_dft_1d(num_samples, in, out, FFTW_FORWARD, FFTW_ESTIMATE);
    fftw_execute(p);

    // Compute magnitude of FFT output
    vector<double> range_profile(num_samples);
    for (int i = 0; i < num_samples; ++i) {
        double real = out[i][0];
        double imag = out[i][1];
        range_profile[i] = sqrt(real*real + imag*imag);
    }

    // Normalize
    double max_val = *max_element(range_profile.begin(), range_profile.end());
    for (int i = 0; i < num_samples; ++i) {
        range_profile[i] /= max_val;
    }

    // Clean up FFTW
    fftw_destroy_plan(p);
    fftw_free(in);
    fftw_free(out);

    return range_profile;
}

File: Range/range_processing.m
function range_profile = range_processing(rx_signal, num_samples)
    % RANGE_PROCESSING Perform range processing on received signal
    %   range_profile = RANGE_PROCESSING(rx_signal, num_samples)
    %
    %   Inputs:
    %       rx_signal: Complex received signal (1 x num_samples)
    %       num_samples: Number of samples in the received signal
    %
    %   Output:
    %       range_profile: Magnitude of the FFT of the windowed signal

    % Apply Hann window
    hann_window = 0.5 * (1 - cos(2 * pi * (0:num_samples-1) / (num_samples-1)));
    windowed_signal = hann_window .* rx_signal;

    % Perform FFT
    fft_result = fft(windowed_signal, num_samples);

    % Compute magnitude of FFT output
    range_profile = abs(fft_result);

    % Normalize range profile
    range_profile = range_profile / max(range_profile);
end

% Example usage
num_samples = 1024;

% Generate a sample received signal (replace with actual radar data)
t = (0:num_samples-1) / num_samples;
rx_signal = exp(1j * 2 * pi * 10 * t); % Example: 10 Hz complex sinusoid

% Process the range data
range_profile = range_processing(rx_signal, num_samples);

% Plot the range profile
figure;
plot(range_profile);
xlabel('Range Bin');
ylabel('Magnitude');
title('Range Profile');
grid on;

% Display the first few values of the range profile
disp('Range Profile (first 10 values):');
disp(range_profile(1:10));
